sedarwin7/src/darwin/adv_cmds/ps.tproj/ps.1
.\" Copyright (c) 1980, 1990, 1991, 1993, 1994
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)ps.1	8.3 (Berkeley) 4/18/94
.\"	$FreeBSD: ps.1,v 1.19 1998/06/04 07:02:59 jkoshy Exp $
.\"
.Dd April 18, 1994
.Dt PS 1
.Os BSD 4
.Sh NAME
.Nm ps
.Nd process status
.Sh SYNOPSIS
.Nm ps
.Op Fl aCcefhjlMmrSTuvwxZ
.Op Fl O Ar fmt
.Op Fl o Ar fmt
.Op Fl p Ar pid
.Op Fl t Ar tty
.Op Fl U Ar username
.Nm ps
.Op Fl L
.Sh DESCRIPTION
.Nm Ps
displays a header line followed by lines containing information about your
processes that have controlling terminals.
This information is sorted by controlling terminal, then by process
.Tn ID .
.Pp
The information displayed is selected based on a set of keywords (see the
.Fl L
.Fl O
and
.Fl o
options).
The default output format includes, for each process, the process'
.Tn ID ,
controlling terminal, cpu time (including both user and system time),
state, and associated command.
.Pp
The process file system (see
.Xr procfs 5
\&) should be mounted when
.Nm
is executed, otherwise not all information will be available.
.Pp
The options are as follows:
.Bl -tag -width indent
.It Fl a
Display information about other users' processes as well as your own.
.It Fl c
Change the ``command'' column output to just contain the executable name,
rather than the full command line.
.It Fl C
Change the way the cpu percentage is calculated by using a ``raw''
cpu calculation that ignores ``resident'' time (this normally has
no effect).
.It Fl e
Display the environment as well.
.It Fl f
Show commandline and environment information about swapped out processes.
This option is honored only if the uid of the user is 0.
.It Fl h
Repeat the information header as often as necessary to guarantee one
header per page of information.
.It Fl j
Print information associated with the following keywords:
user, pid, ppid, pgid, sess, jobc, state, tt, time and command.
.It Fl L
List the set of available keywords.
.It Fl l
Display information associated with the following keywords:
uid, pid, ppid, cpu, pri, nice, vsz, rss, wchan, state, tt, time
and command.
.It Fl M
Print the threads corresponding to each task.
.It Fl m
Sort by memory usage, instead of by process
.Tn ID .
.It Fl O
Add the information associated with the space or comma separated list
of keywords specified, after the process
.Tn ID ,
in the default information
display.
Keywords may be appended with an equals (``='') sign and a string.
This causes the printed header to use the specified string instead of
the standard header.
.It Fl o
Display information associated with the space or comma separated list
of keywords specified.
Keywords may be appended with an equals (``='') sign and a string.
This causes the printed header to use the specified string instead of
the standard header.
.It Fl p
Display information associated with the specified process
.Tn ID .
.It Fl r
Sort by current cpu usage, instead of by process
.Tn ID .
.It Fl S
Change the way the process time is calculated by summing all exited
children to their parent process.
.It Fl T
Display information about processes attached to the device associated
with the standard input.
.It Fl t
Display information about processes attached to the specified terminal
device.
.It Fl U
Display the processes belonging to the specified
.Tn username .
.It Fl u
Display information associated with the following keywords:
user, pid, %cpu, %mem, vsz, rss, tt, state, start, time and command.
The
.Fl u
option implies the
.Fl r
option.
.It Fl v
Display information associated with the following keywords:
pid, state, time, sl, re, pagein, vsz, rss, lim, tsiz,
%cpu, %mem and command.
The
.Fl v
option implies the
.Fl m
option.
.It Fl w
Use 132 columns to display information, instead of the default which
is your window size.
If the
.Fl w
option is specified more than once,
.Nm
will use as many columns as necessary without regard for your window size.
.It Fl x
Display information about processes without controlling terminals.
.It Fl Z
Add label to the list of keywords for which
.Nm
will display information.  This will display label information for
loaded MAC Framework policies, configured in /etc/mac.conf.
.El
.Pp
A complete list of the available keywords are listed below.
Some of these keywords are further specified as follows:
.Bl -tag -width indent
.It %cpu
The cpu utilization of the process; this is a decaying average over up to
a minute of previous (real) time.
Since the time base over which this is computed varies (since processes may
be very young) it is possible for the sum of all
.Tn \&%CPU
fields to exceed 100%.
.It %mem
The percentage of real memory used by this process.
.It flags
The flags associated with the process as in
the include file
.Aq Pa sys/proc.h :
.Bl -column P_NOCLDSTOP P_NOCLDSTOP
.It Dv "P_ADVLOCK" Ta No "0x00001	Process may hold a POSIX advisory lock"
.It Dv "P_CONTROLT" Ta No "0x00002	Has a controlling terminal"
.It Dv "P_INMEM" Ta No "0x00004		Loaded into memory"
.It Dv "P_NOCLDSTOP" Ta No "0x00008	No SIGCHLD when children stop"
.It Dv "P_PPWAIT" Ta No "0x00010	Parent is waiting for child to exec/exit"
.It Dv "P_PROFIL" Ta No "0x00020	Has started profiling"
.It Dv "P_SELECT" Ta No "0x00040	Selecting; wakeup/waiting danger"
.It Dv "P_SINTR" Ta No "0x00080		Sleep is interruptible"
.It Dv "P_SUGID" Ta No "0x00100		Had set id privileges since last exec"
.It Dv "P_SYSTEM" Ta No "0x00200	System proc: no sigs, stats or swapping"
.It Dv "P_TIMEOUT" Ta No "0x00400	Timing out during sleep"
.It Dv "P_TRACED" Ta No "0x00800	Debugged process being traced"
.It Dv "P_WAITED" Ta No "0x01000	Debugging process has waited for child"
.It Dv "P_WEXIT" Ta No "0x02000		Working on exiting"
.It Dv "P_EXEC" Ta No "0x04000		Process called exec"
.It Dv "P_NOSWAP" Ta No "0x08000	Another flag to prevent swap out"
.It Dv "P_PHYSIO" Ta No "0x10000	Doing physical I/O"
.It Dv "P_OWEUPC" Ta No "0x20000	Owe process an addupc() call at next ast"
.It Dv "P_SWAPPING" Ta No "0x40000	Process is being swapped"
.El
.It label
The MAC label of the process.
.It lim
The soft limit on memory used, specified via a call to
.Xr setrlimit 2 .
.It lstart
The exact time the command started, using the ``%c'' format described in
.Xr strftime 3 .
.It nice
The process scheduling increment (see
.Xr setpriority 2 ) .
.It rss
the real memory (resident set) size of the process (in 1024 byte units).
.It start
The time the command started.
If the command started less than 24 hours ago, the start time is
displayed using the ``%l:ps.1p'' format described in
.Xr strftime 3 .
If the command started less than 7 days ago, the start time is
displayed using the ``%a6.15p'' format.
Otherwise, the start time is displayed using the ``%e%b%y'' format.
.It state
The state is given by a sequence of letters, for example,
.Dq Tn RWNA .
The first letter indicates the run state of the process:
.Pp
.Bl -tag -width indent -compact
.It D
Marks a process in disk (or other short term, uninterruptible) wait.
.It I
Marks a process that is idle (sleeping for longer than about 20 seconds).
.It R
Marks a runnable process.
.It S
Marks a process that is sleeping for less than about 20 seconds.
.It T
Marks a stopped process.
.It Z
Marks a dead process (a ``zombie'').
.El
.Pp
Additional characters after these, if any, indicate additional state
information:
.Pp
.Bl -tag -width indent -compact
.It +
The process is in the foreground process group of its control terminal.
.It <
The process has raised
.Tn CPU
scheduling priority.
.It >
The process has specified a soft limit on memory requirements and is
currently exceeding that limit; such a process is (necessarily) not
swapped.
.It A
the process has asked for random page replacement
.Pf ( Dv VA_ANOM ,
from
.Xr vadvise 2 ,
for example,
.Xr lisp 1
in a garbage collect).
.It E
The process is trying to exit.
.It L
The process has pages locked in core (for example, for raw
.Tn I/O ) .
.It N
The process has reduced
.Tn CPU
scheduling priority (see
.Xr setpriority 2 ) .
.It S
The process has asked for
.Tn FIFO
page replacement
.Pf ( Dv VA_SEQL ,
from
.Xr vadvise 2 ,
for example, a large image processing program using virtual memory to
sequentially address voluminous data).
.It s
The process is a session leader.
.It V
The process is suspended during a
.Xr vfork .
.It W
The process is swapped out.
.It X
The process is being traced or debugged.
.El
.It tt
An abbreviation for the pathname of the controlling terminal, if any.
The abbreviation consists of the three letters following
.Pa /dev/tty ,
or, for the console, ``con''.
This is followed by a ``-'' if the process can no longer reach that
controlling terminal (i.e., it has been revoked).
.It wchan
The event (an address in the system) on which a process waits.
When printed numerically, the initial part of the address is
trimmed off and the result is printed in hex, for example, 0x80324000 prints
as 324000.
.El
.Pp
When printing using the command keyword, a process that has exited and
has a parent that has not yet waited for the process (in other words, a zombie)
is listed as ``<defunct>'', and a process which is blocked while trying
to exit is listed as ``<exiting>''.
.Nm Ps
makes an educated guess as to the file name and arguments given when the
process was created by examining memory or the swap area.
The method is inherently somewhat unreliable and in any event a process
is entitled to destroy this information, so the names cannot be depended
on too much.
The ucomm (accounting) keyword can, however, be depended on.
.Sh KEYWORDS
The following is a complete list of the available keywords and their
meanings.
Several of them have aliases (keywords which are synonyms).
.Pp
.Bl -tag -width sigignore -compact
.It %cpu
percentage cpu usage (alias pcpu)
.It %mem
percentage memory usage (alias pmem)
.It acflag
accounting flag (alias acflg)
.It command
command and arguments
.It cpu
short-term cpu usage factor (for scheduling)
.It flags
the process flags, in hexadecimal (alias f)
.It gid
the effective gid
.It inblk
total blocks read (alias inblock)
.It jobc
job control count
.It ktrace
tracing flags
.It ktracep
tracing vnode
.It label
MAC label
.It lim
memoryuse limit
.It logname
login name of user who started the process

sedarwin7/src/darwin/adv_cmds/ps.tproj/ps.c
/*-
 * Copyright (c) 2005 SPARTA, Inc.
 *      
 * Copyright (c) 1990, 1993, 1994
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifndef lint
static char const copyright[] =
"@(#) Copyright (c) 1990, 1993, 1994\n\
	The Regents of the University of California.  All rights reserved.\n";
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)ps.c	8.4 (Berkeley) 4/2/94";
#endif
static const char rcsid[] =
	"$FreeBSD: ps.c,v 1.25 1998/06/30 21:34:14 phk Exp $";
#endif /* not lint */

#include <sys/param.h>
#include <sys/user.h>
#include <sys/time.h>
#include <sys/resource.h>
#include <sys/stat.h>
#include <sys/ioctl.h>
#include <sys/sysctl.h>
#include <sys/syslog.h>

#include <ctype.h>
#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <kvm.h>
#include <limits.h>
#include <nlist.h>
#include <paths.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <locale.h>
#include <pwd.h>

#include "ps.h"

KINFO *kinfo;
struct varent *vhead, *vtail;

int	eval;			/* exit value */
int	cflag;			/* -c */
int	rawcpu;			/* -C */
int	sumrusage;		/* -S */
int	termwidth;		/* width of screen (0 == infinity) */
int	totwidth;		/* calculated width of requested variables */

static int needuser, needcomm, needenv;
#if defined(LAZY_PS)
static int forceuread=0;
#else
static int forceuread=1;
#endif

enum sort { DEFAULT, SORTMEM, SORTCPU } sortby = DEFAULT;

static char	*fmt __P((char **(*)(kvm_t *, const struct kinfo_proc *, int),
		    KINFO *, char *, int));
static char	*kludge_oldps_options __P((char *));
static int	 pscomp __P((const void *, const void *));
static void	 saveuser __P((KINFO *));
static void	 scanvars __P((void));
static void	 dynsizevars __P((KINFO *));
static void	 sizevars __P((void));
static void	 usage __P((void));

char dfmt[] = "pid tt state time command";
char jfmt[] = "user pid ppid pgid sess jobc state tt time command";
char lfmt[] = "uid pid ppid cpu pri nice vsz rss wchan state tt time command";
char   o1[] = "pid";
char   o2[] = "tt state time command";
char ufmt[] = "user pid %cpu %mem vsz rss tt state start time command";
char mfmt[] = "user pid tt %cpu state  pri stime utime command";
char vfmt[] = "pid state time sl re pagein vsz rss lim tsiz %cpu %mem command";
char Zfmt[] = "label";

int mflg = 0; /* if -M option to display all mach threads */
int print_thread_num=0;
int print_all_thread=0;
kvm_t *kd;
int eflg=0;
int
main(argc, argv)
	int argc;
	char *argv[];
{
	struct kinfo_proc *kp, *kprocbuf;
	struct varent *vent;
	struct winsize ws;
	struct passwd *pwd;
	dev_t ttydev;
	pid_t pid;
	uid_t uid;
	pid_t lcid;
	int all, ch, flag, i, j, fmt, lineno, nentries, dropgid;
	int prtheader, wflag, what, xflg;
	char *nlistf, *memf, *swapf, errbuf[_POSIX2_LINE_MAX];
	int mib[4] = { CTL_KERN, KERN_PROC, KERN_PROC_ALL, 0 };
	size_t bufSize = 0;
	size_t orig_bufSize = 0;
	int local_error=0;
	int retry_count = 0;

	(void) setlocale(LC_ALL, "");

	if ((ioctl(STDOUT_FILENO, TIOCGWINSZ, (char *)&ws) == -1 &&
	     ioctl(STDERR_FILENO, TIOCGWINSZ, (char *)&ws) == -1 &&
	     ioctl(STDIN_FILENO,  TIOCGWINSZ, (char *)&ws) == -1) ||
	     ws.ws_col == 0)
		termwidth = 79;
	else
		termwidth = ws.ws_col - 1;

	if( argc > 1 )
		argv[1] = kludge_oldps_options(argv[1]);

	all = fmt = prtheader = wflag = xflg = 0;
	pid = -1;
	uid = (uid_t) -1;
	lcid = -1;
	ttydev = NODEV;
	dropgid = 0;
	memf = nlistf = swapf = _PATH_DEVNULL;
	memf = _PATH_MEM;
	while ((ch = getopt(argc, argv,
#if defined(LAZY_PS)
	    "aCcefghjLlM:mN:O:o:p:rSTt:U:uvW:wxZ")) != -1)
#else
	    "aCceghjLlMmO:o:p:rSTt:U:uvwxZ")) != -1)
#endif
		switch((char)ch) {
		case 'a':
			all = 1;
			break;
		case 'C':
			rawcpu = 1;
			break;
		case 'c':
			cflag = 1;
			break;
		case 'e':			/* XXX set ufmt */
			needenv = 1;
			eflg = 1;
			break;
		case 'g':
			break;			/* no-op */
		case 'h':
			prtheader = ws.ws_row > 5 ? ws.ws_row : 22;
			break;
		case 'j':
			parsefmt(jfmt);
			fmt = 1;
			jfmt[0] = '\0';
			break;
		case 'L':
			showkey();
			exit(0);
		case 'l':
			parsefmt(lfmt);
			fmt = 1;
			lfmt[0] = '\0';
			break;
		case 'M':
#if 0
			memf = optarg;
			dropgid = 1;
#else
			parsefmt(mfmt);
			fmt = 1;
			mfmt[0] = '\0';
			mflg  = 1;
#endif /* 0 */
			break;
		case 'm':
			sortby = SORTMEM;
			break;
		case 'N':
			nlistf = optarg;
			dropgid = 1;
			break;
		case 'O':
			parsefmt(o1);
			parsefmt(optarg);
			parsefmt(o2);
			o1[0] = o2[0] = '\0';
			fmt = 1;
			break;
		case 'o':
			parsefmt(optarg);
			fmt = 1;
			break;
#if defined(LAZY_PS)
		case 'f':
			if (getuid() == 0 || getgid() == 0)
			    forceuread = 1;
			break;
#endif
		case 'p':
		        /* Given more than one process requested with -p
			   or pid, ps will return information on only
			   the first requested process as is done by FreeBSD */
		        if (pid != -1) break;

			pid = atol(optarg);
			xflg = 1;
			break;
		case 'r':
			sortby = SORTCPU;
			break;
		case 'S':
			sumrusage = 1;
			break;
		case 'T':
			if ((optarg = ttyname(STDIN_FILENO)) == NULL)
				errx(1, "stdin: not a terminal");
			/* FALLTHROUGH */
		case 't': {
			struct stat sb;
			char *ttypath, pathbuf[MAXPATHLEN];

			if (strcmp(optarg, "co") == 0)
				ttypath = _PATH_CONSOLE;
			else if (*optarg != '/')
				(void)snprintf(ttypath = pathbuf,
				    sizeof(pathbuf), "%s%s", _PATH_TTY, optarg);
			else
				ttypath = optarg;
			if (stat(ttypath, &sb) == -1)
				err(1, "%s", ttypath);
			if (!S_ISCHR(sb.st_mode))
				errx(1, "%s: not a terminal", ttypath);
			ttydev = sb.st_rdev;
			break;
		}
		case 'U':
			pwd = getpwnam(optarg);
			if (pwd == NULL)
				errx(1, "%s: no such user", optarg);
			uid = pwd->pw_uid;
			endpwent();
			xflg++;		/* XXX: intuitive? */
			break;
		case 'u':
			parsefmt(ufmt);
			sortby = SORTCPU;
			fmt = 1;
			ufmt[0] = '\0';
			break;
		case 'v':
			parsefmt(vfmt);
			sortby = SORTMEM;
			fmt = 1;
			vfmt[0] = '\0';
			break;
		case 'W':
			swapf = optarg;
			dropgid = 1;
			break;
		case 'w':
			if (wflag)
				termwidth = UNLIMITED;
			else if (termwidth < 131)
				termwidth = 131;
			wflag++;
			break;
		case 'x':
			xflg = 1;
			break;
		case 'Z':
			parsefmt(Zfmt);
			Zfmt[0] = '\0';
			break;
		case '?':
		default:
			usage();
		}
	argc -= optind;
	argv += optind;

#define	BACKWARD_COMPATIBILITY
#ifdef	BACKWARD_COMPATIBILITY
	if (*argv) {
		nlistf = *argv;
		if (*++argv) {
			memf = *argv;
			if (*++argv)
				swapf = *argv;
		}
	}
#endif
	/*
	 * Discard setgid privileges if not the running kernel so that bad
	 * guys can't print interesting stuff from kernel memory.
	 */
	if (dropgid) {
		setgid(getgid());
		setuid(getuid());
	}
#if FIXME 
	kd = kvm_openfiles(nlistf, NULL, swapf, O_RDONLY, errbuf);
	if (kd == 0)
		errx(1, "%s", errbuf);
#endif /* FIXME */

	if (!fmt)
           parsefmt(dfmt);

	/* XXX - should be cleaner */
	if (!all && ttydev == NODEV && pid == -1 && uid == (uid_t)-1)
		uid = getuid();

	/*
	 * scan requested variables, noting what structures are needed,
	 * and adjusting header widths as appropiate.
	 */
	scanvars();

	/*
	 * get proc list
	 */
	if (uid != (uid_t) -1) {
		what = KERN_PROC_UID;
		flag = uid;
	} else if (ttydev != NODEV) {
		what = KERN_PROC_TTY;
		flag = ttydev;
	} else if (pid != -1) {
		what = KERN_PROC_PID;
		flag = pid;
	} else if (lcid != -1) {
		what = KERN_PROC_LCID;
		flag = lcid;
	} else {
		what = KERN_PROC_ALL;
		flag = 0;
	}
	/*
	 * select procs
	 */
#if FIXME
	if ((kp = kvm_getprocs(kd, what, flag, &nentries)) == 0)
		errx(1, "%s", kvm_geterr(kd));
#else /* FIXME */

    mib[0] = CTL_KERN;
    mib[1] = KERN_PROC;
    mib[2] = what;
    mib[3] = flag;

    if (sysctl(mib, 4, NULL, &bufSize, NULL, 0) < 0) {
        perror("Failure calling sysctl");
        return 0;
    }

    kprocbuf= kp = (struct kinfo_proc *)malloc(bufSize);

    retry_count = 0;
    orig_bufSize = bufSize;
   for(retry_count=0; ; retry_count++) {
    /* retry for transient errors due to load in the system */
    local_error = 0;
    bufSize = orig_bufSize;
    if ((local_error = sysctl(mib, 4, kp, &bufSize, NULL, 0)) < 0) {
	if (retry_count < 1000) {
		/* 1 sec back off */
		sleep(1);
		continue;
	}
        perror("Failure calling sysctl");
        return 0;
    } else if (local_error == 0) {
	break;
    }
    /* 1 sec back off */
    sleep(1);
   }

    /* This has to be after the second sysctl since the bufSize
       may have changed.  */
    nentries = bufSize/ sizeof(struct kinfo_proc);

#endif /* FIXME */
	if ((kinfo = malloc(nentries * sizeof(KINFO))) == NULL)
		err(1, NULL);
	memset(kinfo, 0, (nentries * sizeof(*kinfo)));
	for (i = nentries; --i >= 0; ++kp) {
		kinfo[i].ki_p = kp;
#if 1
		get_task_info(&kinfo[i]);
#endif /* 1 */
		if (needuser)
			saveuser(&kinfo[i]);
		dynsizevars(&kinfo[i]);
	}
	sizevars();

	/*
	 * print header
	 */
	printheader();
	if (nentries == 0)
		exit(0);
	/*
	 * sort proc list
	 */
	qsort(kinfo, nentries, sizeof(KINFO), pscomp);
	/*
	 * for each proc, call each variable output function.
	 */
	for (i = lineno = 0; i < nentries; i++) {
		if(kinfo[i].invalid_tinfo || kinfo[i].invalid_thinfo)
			continue;
		if (KI_PROC(&kinfo[i])->p_pid == 0)
			continue;
		if (xflg == 0 && (KI_EPROC(&kinfo[i])->e_tdev == NODEV ||
		    (KI_PROC(&kinfo[i])->p_flag & P_CONTROLT ) == 0))
			continue;
		if(mflg) {
			print_all_thread = 1;
			for(j=0; j < kinfo[i].thread_count; j++) {
			print_thread_num = j;
			for (vent = vhead; vent; vent = vent->next) {
				(vent->var->oproc)(&kinfo[i], vent);
				if (vent->next != NULL)
					(void)putchar(' ');
			}

			(void)putchar('\n');
			if (prtheader && lineno++ == prtheader - 4) {
				(void)putchar('\n');
				printheader();
				lineno = 0;
			}
			}
			print_all_thread = 0;
		} else {
			for (vent = vhead; vent; vent = vent->next) {
				(vent->var->oproc)(&kinfo[i], vent);
				if (vent->next != NULL)
					(void)putchar(' ');
			}

			(void)putchar('\n');
			if (prtheader && lineno++ == prtheader - 4) {
				(void)putchar('\n');
				printheader();
				lineno = 0;
			}
		}
	}
cleanup:
	for (i = 0; i < nentries; i++) {
		if(kinfo[i].thread_count)
			free(kinfo[i].thval);	
	}
	free(kprocbuf);
	free(kinfo);
	exit(eval);
}

static void
scanvars() 
{
	struct varent *vent;
	VAR *v;

        struct varent *prev, *curr, *command;

        /* First we make sure that COMMAND is the last element
	   in the list */
        command = NULL;
        prev = NULL;
        curr = vhead;
        while (curr) {
            if (!strcmp(curr->var->header, "COMMAND")) {
                if (prev == NULL || (curr->next == NULL)) {
                    break;
                }
                else {
                    command = curr;
                    prev->next = curr->next;
                }
            }
            prev = curr;
            curr = curr->next;
        }

        if (command) {
            prev->next = command;
            command->next = NULL;
        }
    
        /* Now we set the spacing information */
	for (vent = vhead; vent; vent = vent->next) {
		v = vent->var;
		if (v->flag & DSIZ) {
			v->dwidth = v->width;
			v->width = 0;
		}
		if (v->flag & USER)
			needuser = 1;
		if (v->flag & COMM)
			needcomm = 1;
	}
}

static void
dynsizevars(ki)
	KINFO *ki;
{
	struct varent *vent;
	VAR *v;
	int i;

	for (vent = vhead; vent; vent = vent->next) {
		v = vent->var;
		if (!(v->flag & DSIZ))
			continue;
		i = (v->sproc)( ki);
		if (v->width < i)
			v->width = i;
		if (v->width > v->dwidth)
			v->width = v->dwidth;
	}
}

static void
sizevars()
{
	struct varent *vent;
	VAR *v;
	int i;

	for (vent = vhead; vent; vent = vent->next) {
		v = vent->var;
		i = strlen(v->header);
		if (v->width < i)
			v->width = i;
		totwidth += v->width + 1;	/* +1 for space */
	}
	totwidth--;
}

static char *
fmt(fn, ki, comm, maxlen)
	char **(*fn) __P((kvm_t *, const struct kinfo_proc *, int));
	KINFO *ki;
	char *comm;
	int maxlen;
{
	char *s;

	if ((s =
	    fmt_argv((*fn)(kd, ki->ki_p, termwidth), comm, maxlen)) == NULL)
		err(1, NULL);
	return (s);
}

#define UREADOK(ki)	(forceuread || (KI_PROC(ki)->p_flag & P_INMEM))

static void
saveuser(ki)
	KINFO *ki;
{
	struct pstats pstats;
	struct usave *usp;
#if FIXME
	struct user *u_addr = (struct user *)USRSTACK;
#else /* FIXME */
	struct user *u_addr = (struct user *)0;
#endif /* FIXME */

	usp = &ki->ki_u;
#if FIXME
	if (UREADOK(ki) && kvm_uread(kd, KI_PROC(ki), (unsigned long)&u_addr->u_stats,
	    (char *)&pstats, sizeof(pstats)) == sizeof(pstats)) 
	{
		/*
		 * The u-area might be swapped out, and we can't get
		 * at it because we have a crashdump and no swap.
		 * If it's here fill in these fields, otherwise, just
		 * leave them 0.
		 */
		usp->u_start = pstats.p_start;
		usp->u_ru = pstats.p_ru;
		usp->u_cru = pstats.p_cru;
		usp->u_valid = 1;
	} else
		usp->u_valid = 0;
#else /* FIXME */
		usp->u_valid = 0;
#endif /* FIXME */
	/*
	 * save arguments if needed
	 */
#if FIXME
	if (needcomm && UREADOK(ki)) {
	    ki->ki_args = fmt(kvm_getargv, ki, KI_PROC(ki)->p_comm,
		MAXCOMLEN);
	} else if (needcomm) {
	    ki->ki_args = malloc(strlen(KI_PROC(ki)->p_comm) + 3);
	    sprintf(ki->ki_args, "(%s)", KI_PROC(ki)->p_comm);
    } else {
	    ki->ki_args = NULL;
    }
#else /* FIXME */
	    ki->ki_args = malloc(strlen(KI_PROC(ki)->p_comm) + 3);
	    sprintf(ki->ki_args, "%s", KI_PROC(ki)->p_comm);
	    //ki->ki_args = malloc(10);
	    //strcpy(ki->ki_args, "()");
#endif /* FIXME */
#if FIXME
   if (needenv && UREADOK(ki)) {
	    ki->ki_env = fmt(kvm_getenvv, ki, (char *)NULL, 0);
    } else if (needenv) {
	    ki->ki_env = malloc(3);
	    strcpy(ki->ki_env, "()");
    } else {
	    ki->ki_env = NULL;
    }
#else /* FIXME */
	    ki->ki_env = malloc(10);
	    strcpy(ki->ki_env, "");
#endif /* FIXME */
}

static int
pscomp(a, b)
	const void *a, *b;
{
	int i;
#if FIXME
#define VSIZE(k) (KI_EPROC(k)->e_vm.vm_dsize + KI_EPROC(k)->e_vm.vm_ssize + \
		  KI_EPROC(k)->e_vm.vm_tsize)
#else /* FIXME */
#define VSIZE(k) ((k)->tasks_info.resident_size)

#endif /* FIXME */

	if (sortby == SORTCPU)
		return (getpcpu((KINFO *)b) - getpcpu((KINFO *)a));
	if (sortby == SORTMEM)
		return (VSIZE((KINFO *)b) - VSIZE((KINFO *)a));
	i =  KI_EPROC((KINFO *)a)->e_tdev - KI_EPROC((KINFO *)b)->e_tdev;
	if (i == 0)
		i = KI_PROC((KINFO *)a)->p_pid - KI_PROC((KINFO *)b)->p_pid;
	return (i);
}

/*
 * ICK (all for getopt), would rather hide the ugliness
 * here than taint the main code.
 *
 *  ps foo -> ps -foo
 *  ps 34 -> ps -p34
 *
 * The old convention that 't' with no trailing tty arg means the users
 * tty, is only supported if argv[1] doesn't begin with a '-'.  This same
 * feature is available with the option 'T', which takes no argument.
 */
static char *
kludge_oldps_options(s)
	char *s;
{
	size_t len;
	char *newopts, *ns, *cp;

	len = strlen(s);
	if ((newopts = ns = malloc(len + 2)) == NULL)
		err(1, NULL);
	/*
	 * options begin with '-'
	 */
	if (*s != '-')
		*ns++ = '-';	/* add option flag */
	/*
	 * gaze to end of argv[1]
	 */
	cp = s + len - 1;
	/*
	 * if last letter is a 't' flag with no argument (in the context
	 * of the oldps options -- option string NOT starting with a '-' --
	 * then convert to 'T' (meaning *this* terminal, i.e. ttyname(0)).
	 *
	 * However, if a flag accepting a string argument is found in the
	 * option string, the remainder of the string is the argument to
	 * that flag; do not modify that argument.
	 */
	if (strcspn(s, "MNOoUW") == len && *cp == 't' && *s != '-')
		*cp = 'T';
	else {
		/*
		 * otherwise check for trailing number, which *may* be a
		 * pid.
		 */
		while (cp >= s && isdigit(*cp))
			--cp;
	}
	cp++;
	memmove(ns, s, (size_t)(cp - s));	/* copy up to trailing number */
	ns += cp - s;
	/*
	 * if there's a trailing number, and not a preceding 'p' (pid) or
	 * 't' (tty) flag, then assume it's a pid and insert a 'p' flag.
	 */
	if (isdigit(*cp) &&
	    (cp == s || (cp[-1] != 't' && cp[-1] != 'p')) &&
	    (cp - 1 == s || cp[-2] != 't'))
		*ns++ = 'p';
	(void)strcpy(ns, cp);		/* and append the number */

	return (newopts);
}

static void
usage()
{

	(void)fprintf(stderr, "%s\n%s\n%s\n",
	    "usage: ps [-aChjlmMrSTuvwxZ] [-O|o fmt] [-p pid] [-t tty] [-U user]",
	    "          [-N system] [-W swap]",
	    "       ps [-L]");
	exit(1);
}
